### PR TITLE
[react-dom] fix signature of findDOMNode

### LINKED ISSUE #24167: ReactDOM.findDOMNode method has incorrect signature.
- [ ] I tried using the `@types/react-dom` package and had problems.
- [ ] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [ ] I have a question that is inappropriate for StackOverflow.
- [ ] Mention the authors so they can respond.
  - Authors: @RiskMatch, @MartynasZilinskas

Signature should match this https://github.com/facebook/react/blob/master/packages/react-dom/src/client/ReactDOM.js#L1124

Tested with: 
- `react-dom`: 16.2.0
- `@types/react-dom`: 16.0.4

### PR DESCRIPTION
**Pull request by Pajn**
fixes #24167

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/24167
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by Pajn** [comment]
I did have to remove the explicit type information in the react tests as react does not have `lib.dom.d.ts` which provides `Text`. What to do with all other packages that depends on the invalid types and likely will crash in runtime?

**Comment by Pajn** [comment]
The victory error is apparently due to the victory test being wrong, #24769

File: types/react-dom/index.d.ts [hunk 1]
```diff
@@ -16,7 +16,7 @@ import {
     DOMAttributes, DOMElement, ReactNode, ReactPortal
 } from 'react';
 
-export function findDOMNode(instance: ReactInstance): Element;
+export function findDOMNode(instance: ReactInstance): Element | null | Text;
```
Comment (MartynasZilinskas) [inline types/react-dom/index.d.ts]: What is `Text`?

Comment (Kovensky) [reply types/react-dom/index.d.ts]: I thought this was intended to be `TextNode`, but `findDOMNode` on a component that only returns a string actually produces an object that just holds a reference to the Fiber instance.

```js
ReactDOM.render(
  React.createElement(class extends React.Component {
    componentDidMount () {
      console.log(ReactDOM.findDOMNode(this))
    }
    render () {
      return 'test'
    }
  }),
  document.body.appendChild(document.createElement('div'))
)
```

image

Comment (Kovensky) [reply types/react-dom/index.d.ts]: Scratch all of the above; that's just the codesandbox console not printing the `@@toStringTag` or constructor name for instances. That's no ordinary object, it's an instance of the `Text` DOM constructor.

**Review by Jessidhia — APPROVED** [review]
This looks right to me after investigating but I get the nagging feeling this just reduces to `(Node & ChildNode) | null`*.

*: the only missing key that's in `Element|Text` but not in `Node&ChildNode` is `assignedSlot`

**Comment by Kovensky** [comment]
I just meant that those 2 types have so little in common (DOM-wise) that they almost reduce to a mere `Node & ChildNode` w.r.t. which keys the compiler will let you access without refinement. The type is correct, though, in that it enumerates the types that may be returned.

**Comment by wu38607** [comment]
`findDOMNode(component).getBoundingClientRect()`

>  TS2339: Property 'getBoundingClientRect' does not exist on type 'Element | Text'.
[0]   Property 'getBoundingClientRect' does not exist on type 'Text'.

Should I use type assertions?
 `(findDOMNode(component) as Element).getBoundingClientRect()` ?

**Comment by Strate** [comment]
@wu38607 seems that you should use `instanceof` check:
```ts
const node = findDOMNode(component);
if (node instanceof Element) {
  node.getBoundingClientRect()
}
```

File: types/react-dom/index.d.ts (hunk 1 above)
Comment (zkeyword) [reply types/react-dom/index.d.ts]: node_modules/@types/react-dom/index.d.ts(19,72): error TS2304: Cannot find name 'Text'

**Pull request merged**